File: .indio-backups/server-ts-20260507-002053/src/adapters/netease/radio.ts
import type { Track } from "@indio/contracts";

export interface RadioLibraryTrack extends Track {
  playlistNames: string[];
}

function scoreLibraryTrack(track: RadioLibraryTrack, query: string): number {
  const normalized = query.toLowerCase();
  const haystacks = [
    track.title.toLowerCase(),
    track.artist.toLowerCase(),
    track.album.toLowerCase(),
    ...track.playlistNames.map((playlist) => playlist.toLowerCase())
  ];

  let score = 0;

  for (const haystack of haystacks) {
    if (haystack.includes(normalized)) {
      score += 6;
    }
  }

  if (normalized.includes("focus") || normalized.includes("专注") || normalized.includes("工作")) {
    score += track.playlistNames.some((name) => /专注|focus|工作|coding|study/i.test(name)) ? 5 : 0;
  }

  if (normalized.includes("quiet") || normalized.includes("安静") || normalized.includes("轻")) {
    score += track.playlistNames.some((name) => /安静|轻|night|sleep|chill/i.test(name)) ? 5 : 0;
  }

  if (normalized.includes("morning") || normalized.includes("早")) {
    score += track.playlistNames.some((name) => /早|morning|sunrise/i.test(name)) ? 5 : 0;
  }

  if (normalized.includes("evening") || normalized.includes("晚") || normalized.includes("夜")) {
    score += track.playlistNames.some((name) => /夜|晚|midnight|evening|late/i.test(name)) ? 5 : 0;
  }

  return score;
}

export function trackKey(track: Track): string {
  return track.neteaseId ?? track.id;
}

export function primaryArtistKey(track: Track): string {
  return track.artist
    .split(/[\/,，、&＆]/)[0]
    ?.replace(/\s+/g, " ")
    .trim()
    .toLowerCase() ?? track.artist.toLowerCase();
}

function trackText(track: RadioLibraryTrack): string {
  return [
    track.title,
    track.artist,
    track.album,
    track.mood,
    track.sourcePlaylists?.join(" "),
    track.playlistNames.join(" ")
  ]
    .join(" ")
    .toLowerCase();
}

export function samePrimaryArtist(left: Track | null | undefined, right: Track): boolean {
  if (!left?.artist || !right.artist) {
    return false;
  }

  const leftArtist = primaryArtistKey(left);
  const rightArtist = primaryArtistKey(right);
  return Boolean(leftArtist && rightArtist && leftArtist === rightArtist);
}

function sharedPlaylistCount(track: RadioLibraryTrack, currentTrack?: Track | null): number {
  if (!currentTrack?.sourcePlaylists?.length) {
    return 0;
  }

  const currentPlaylists = new Set(currentTrack.sourcePlaylists);
  return track.playlistNames.filter((name) => currentPlaylists.has(name)).length;
}

export function radioSortLibraryTracks<TTrack extends RadioLibraryTrack>(params: {
  tracks: TTrack[];
  hint: string;
  recentTrackIds: Set<string>;
  currentTrack?: Track | null;
  avoidTrackIds?: Set<string>;
}): TTrack[] {
  const normalizedHint = params.hint.toLowerCase();
  const hintWords = normalizedHint
    .split(/[\s,，。/|·:：;；]+/)
    .map((word) => word.trim())
    .filter((word) => word.length >= 2);

  return [...params.tracks]
    .map((track) => {
      const key = trackKey(track);
      const text = trackText(track);
      const directScore = scoreLibraryTrack(track, params.hint);
      const wordScore = hintWords.reduce((score, word) => score + (text.includes(word) ? 1.2 : 0), 0);
      const currentArtistScore = samePrimaryArtist(params.currentTrack, track) ? Math.random() * 2 - 2.5 : 0;
      const playlistScore = Math.min(3, sharedPlaylistCount(track, params.currentTrack) * 1.5);
      const multiPlaylistScore = Math.min(2, track.playlistNames.length * 0.35);
      const recentPenalty = params.recentTrackIds.has(key) ? 12 : 0;
      const avoidPenalty = params.avoidTrackIds?.has(key) ? 4 : 0;
      const sameTrackPenalty = params.currentTrack && trackKey(params.currentTrack) === key ? 100 : 0;
      const radioJitter = Math.random() * 6;

      return {
        track,
        score:
          directScore +
          wordScore +
          currentArtistScore +
          playlistScore +
          multiPlaylistScore +
          radioJitter -
          recentPenalty -
          avoidPenalty -
          sameTrackPenalty
      };
    })
    .filter((entry) => entry.score > -5)
    .sort((left, right) => right.score - left.score)
    .map((entry) => entry.track);
}

File: agent/adapters/netease/radio.py
from __future__ import annotations

import random
import re

from models import Track


def track_key(track: Track) -> str:
    return track.netease_id or track.id


def primary_artist_key(track: Track) -> str:
    first = re.split(r"[/,，、&＆]", track.artist)[0] if track.artist else ""
    return re.sub(r"\s+", " ", first).strip().lower()


def same_primary_artist(left: Track | None, right: Track) -> bool:
    if not left:
        return False
    return bool(primary_artist_key(left) and primary_artist_key(left) == primary_artist_key(right))


def _score_library_track(track: Track, query: str) -> float:
    normalized = query.lower()
    haystacks = [
        track.title.lower(),
        track.artist.lower(),
        track.album.lower(),
        *(name.lower() for name in track.source_playlists),
    ]
    score = sum(6 for haystack in haystacks if normalized and normalized in haystack)
    playlist_text = " ".join(track.source_playlists)
    if any(word in normalized for word in ("focus", "专注", "工作")):
        score += 5 if re.search(r"专注|focus|工作|coding|study", playlist_text, re.I) else 0
    if any(word in normalized for word in ("quiet", "安静", "轻")):
        score += 5 if re.search(r"安静|轻|night|sleep|chill", playlist_text, re.I) else 0
    if any(word in normalized for word in ("morning", "早")):
        score += 5 if re.search(r"早|morning|sunrise", playlist_text, re.I) else 0
    if any(word in normalized for word in ("evening", "晚", "夜")):
        score += 5 if re.search(r"夜|晚|midnight|evening|late", playlist_text, re.I) else 0
    return score


def radio_sort_library_tracks(
    tracks: list[Track],
    hint: str,
    recent_track_ids: set[str],
    current_track: Track | None = None,
    avoid_track_ids: set[str] | None = None,
) -> list[Track]:
    normalized = hint.lower()
    words = [word.strip() for word in re.split(r"[\s,，。/|·:：;；]+", normalized) if len(word.strip()) >= 2]

    def score(track: Track) -> float:
        text = " ".join(
            [track.title, track.artist, track.album, track.mood, *track.source_playlists]
        ).lower()
        direct = _score_library_track(track, hint)
        word_score = sum(1.2 for word in words if word in text)
        recent_penalty = 12 if track_key(track) in recent_track_ids else 0
        avoid_penalty = 4 if avoid_track_ids and track_key(track) in avoid_track_ids else 0
        same_track_penalty = 100 if current_track and track_key(current_track) == track_key(track) else 0
        same_artist = random.random() * 2 - 2.5 if same_primary_artist(current_track, track) else 0
        playlist_score = min(2, len(track.source_playlists) * 0.35)
        return direct + word_score + same_artist + playlist_score + random.random() * 6 - recent_penalty - avoid_penalty - same_track_penalty

    ranked = [(track, score(track)) for track in tracks]
    return [track for track, value in sorted(ranked, key=lambda item: item[1], reverse=True) if value > -5]
